gem/refactorise.py

```diff
@@ -112,6 +116,175 @@ def product(*args):
             result.add(sum_indices, atomics, rest)
         return result
 
+    def argument_indices_extent(self, factor):
+        """
+        Returns the product of extents of argument indices of :param: factor
+        """
+        if self.argument_indices is None:
+            raise AssertionError("argument_indices property not initialised.")
+        return numpy.product([i.extent for i in set(factor.free_indices).intersection(self.argument_indices)])
+
+    def unique_sum_indices(self):
+        """generator of unique sum indices, together with their original
+        ordering of this MonomialSum
+        """
+        seen = set()
+        for (sum_indices_set, _), (sum_indices, _) in iteritems(self.ordering):
+            if sum_indices_set not in seen:
+                seen.add(sum_indices_set)
+                yield (sum_indices_set, sum_indices)
+
+    def to_expression(self):
+        """
+        Convert MonomialSum object to gem node. Use associate_product() and
+        associate_sum() to promote hoisting in subsequent code generation.
+        ordering ensures deterministic code generation.
+        :return: gem node represented by this MonomialSum object.
+        """
+        indexsums = []  # The result is summation of indexsums
+        monomial_group = OrderedDict()  # (sum_indices_set, sum_indices) -> [(atomics, rest)]
+        # Group monomials according to their sum indices
+        for key, (sum_indices, atomics) in iteritems(self.ordering):
+            sum_indices_set, _ = key
+            rest = self.monomials[key]
+            if not sum_indices_set:
+                indexsums.append(fast_sum_factorise(sum_indices, atomics + (rest,)))
+            else:
+                monomial_group.setdefault((sum_indices_set, sum_indices), []).append((atomics, rest))
+
+        # Form IndexSum's from each monomial group
+        for (_, sum_indices), monomials in iteritems(monomial_group):
+            all_atomics = [m[0] for m in monomials]
+            all_rest = [m[1] for m in monomials]
+            if len(all_atomics) == 1:
+                # Just one term, add to indexsums directly
+                indexsums.append(fast_sum_factorise(sum_indices, all_atomics[0] + (all_rest[0],)))
+            else:
+                # Form products for each monomial
+                products = [associate_product(atomics + (_rest,))[0] for atomics, _rest in zip(all_atomics, all_rest)]
+                indexsums.append(IndexSum(associate_sum(products)[0], sum_indices))
+
+        return associate_sum(indexsums)[0]
```

Comment:
OK, so this can now be a free function in `coffee_mode.py` and can get a better name than `to_expression`.